enterprise/frontend/src/metabase-enterprise/embedding_iframe_sdk_setup/hooks/use-persist-via-user-setting.ts

```diff
@@ -0,0 +1,55 @@
+import { useDebouncedCallback } from "@mantine/hooks";
+import { useEffect, useState } from "react";
+import _ from "underscore";
+
+import { useUserSetting } from "metabase/common/hooks";
+import type { UserSettings } from "metabase-types/api";
+
+type HookProps<K extends keyof UserSettings, S> = {
+  /** Callback to restore the settings from the user setting. */
+  onRestore: (settings: S | null) => void;
+
+  /** The key of the user setting to persist. The setting must be a JSON string. */
+  settingKey: K;
+
+  /** How long to debounce before saving the settings. */
+  debounceMs: number;
+
+  /** Omit these setting keys from being persisted. */
+  omitKeys?: (keyof S)[];
+
+  /** Do not restore the settings yet when `skipRestore` is true. */
+  skipRestore?: boolean;
+};
+
+export function usePersistJsonViaUserSetting<K extends keyof UserSettings, S>({
+  onRestore,
+  settingKey,
+  debounceMs,
+  omitKeys = [],
+  skipRestore,
+}: HookProps<K, S>) {
+  const [isUserSettingsLoaded, setIsUserSettingsLoaded] = useState(false);
+  const [userSettingString, setUserSettingString] = useUserSetting(settingKey);
+
+  const storeSetting = useDebouncedCallback((settings: S) => {
+    const sanitizedSettings = _.omit(settings, omitKeys as string[]);
+
+    setUserSettingString(JSON.stringify(sanitizedSettings) as UserSettings[K]);
+  }, debounceMs);
+
+  useEffect(() => {
+    if (!isUserSettingsLoaded && !skipRestore) {
```

Comment:
We have to wait until we can restore the user settings. In this case, we wait until the recent dashboards are loaded to avoid a race condition.